### ISSUE TITLE
Recursive dependency detection killed the global build macro star.

### ISSUE THREAD

**Issue by back2dos** [HaxeFoundation/haxe#8254]
Minimal example:

```haxe
// Tst.hx
class Tst {
  static function main() {}
  #if macro
  static function use() 
    haxe.macro.Compiler.addGlobalMetadata('', '@:build(Tst.build())', true, true, false);
  
  static function build():Array<haxe.macro.Expr.Field>
    return null;
  #end
}
```

```haxe
// fail.hxml
--macro Tst.use()
--run Tst
```

With latest nightly (19c2778), compiling a 2nd time with no changes yields: 

> Server.ServerError("Infinite loop in Haxe server detected. Probably caused by shadowing a module of the standard library. Make sure shadowed module does not pull macro context.")

**Comment by Gama11** [comment]
Potentially fixed by #8196?

**Comment by RealyUniqueName** [comment]
> Potentially fixed by #8196?

No, this is a different case.

**Comment by Simn** [comment]
Let's add a test for this to tests/server.
